polygeist/lib/Dialect/Polygeist/Analysis/MemoryAccessAnalysis.cpp

```diff
@@ -92,22 +101,28 @@ static SmallVector<Value> computeThreadVector(FunctionOpInterface funcOp,
                    [&](T &op) { return !getIndexValue(op).has_value(); }))
     return {};
 
-  // Order the operations in increasing index value.
-  llvm::sort(getGlobalIdOps, [&](T &op1, T &op2) {
-    return getIndexValue(op1)->slt(*getIndexValue(op2));
-  });
+  // Create a map from operation index to operation result.
+  std::map<unsigned, Value> indexToGlobalOp;
+  for (T &getGlobalIdOp : getGlobalIdOps) {
+    APInt index = *getIndexValue(getGlobalIdOp);
+    indexToGlobalOp[index.getZExtValue()] = getGlobalIdOp.getResult();
+  }
 
-  // Collect the thread values.
-  SmallVector<Value> threadVars;
-  for (T getGlobalIdOp : getGlobalIdOps)
-    threadVars.emplace_back(getGlobalIdOp.getResult());
+  // Collect the thread values/index.
+  SmallVector<std::pair<Value, unsigned>> threadVars;
+  for (unsigned dim = 0; dim < gridDim; ++dim) {
```

Comment:
as dim is the index of the vector, there is no need to have pair as element, it can be `SmallVector<Value>`.